pkg/nimble/autoadv/nimble_autoadv.c

```diff
@@ -87,94 +89,195 @@ int nimble_autoadv_add_field(uint8_t type, const void *data, size_t data_len)
 {
     _gap_cb = cb;
     _gap_cb_arg = cb_arg;
 
-    if (ble_gap_adv_active()) {
-        nimble_autoadv_start();
+    if (_ble_gap_adv_active()) {
+        nimble_autoadv_start(NULL);
     }
 }
 
-void nimble_autoadv_start(void)
+#if MYNEWT_VAL_BLE_EXT_ADV
+static int _get_phy_hci(uint8_t mode)
+{
+    switch (mode) {
+    case NIMBLE_PHY_1M:
+        return BLE_HCI_LE_PHY_1M;
+#if IS_USED(MODULE_NIMBLE_PHY_2MBIT)
+    case NIMBLE_PHY_2M:
+        return BLE_HCI_LE_PHY_2M;
+#endif
+#if IS_USED(MODULE_NIMBLE_PHY_CODED)
+    case NIMBLE_PHY_CODED:
+        return BLE_HCI_LE_PHY_CODED;
+#endif
+    default:
+        return -1;
+    }
+}
+#endif
+
+void nimble_autoadv_start(ble_addr_t *addr)
 {
     int rc;
-    (void) rc;
 
-    rc = ble_gap_adv_stop();
-    assert(rc == BLE_HS_EALREADY || rc == 0);
+    (void)rc;
+
+    nimble_autoadv_stop();
+
+#if MYNEWT_VAL_BLE_EXT_ADV
+    struct ble_gap_ext_adv_params advp = {
+        .connectable = (_cfg.flags & NIMBLE_AUTOADV_FLAG_CONNECTABLE) ? 1 : 0,
+        .scannable = (_cfg.flags & NIMBLE_AUTOADV_FLAG_SCANNABLE) ? 1 : 0,
+        .directed = addr ? 1 : 0,
+        .high_duty_directed = (_cfg.flags & NIMBLE_AUTOADV_FLAG_HD_MODE) ? 1 : 0,
+        .legacy_pdu = (_cfg.flags & NIMBLE_AUTOADV_FLAG_LEGACY) ? 1 : 0,
+        .anonymous = (_cfg.flags & NIMBLE_AUTOADV_FLAG_ANONYMOUS) ? 1 : 0,
+        .include_tx_power = 0,
+        .scan_req_notif = (_cfg.flags & NIMBLE_AUTOADV_FLAG_SCAN_REQ_NOTIF) ? 1 : 0,
+        .itvl_min = BLE_GAP_ADV_ITVL_MS(_cfg.adv_itvl_ms),
+        .itvl_max = BLE_GAP_ADV_ITVL_MS(_cfg.adv_itvl_ms),
+        .channel_map = _cfg.channel_map,
+        .own_addr_type = _cfg.own_addr_type,
+        .peer = *addr,
+        .filter_policy = _cfg.filter_policy,
+        .primary_phy = _get_phy_hci(_cfg.phy),
+        .secondary_phy = _get_phy_hci(_cfg.phy),
+        .tx_power = _cfg.tx_power,
+        .sid = CONFIG_NIMBLE_AUTOADV_INST,
+    };
+
+    /* remove before configure */
+    if (ble_gap_ext_adv_active(CONFIG_NIMBLE_AUTOADV_INST)) {
+        rc = ble_gap_ext_adv_remove(CONFIG_NIMBLE_AUTOADV_INST);
+        assert(rc == 0);
+    }
+    /* configure */
+    rc = ble_gap_ext_adv_configure(CONFIG_NIMBLE_AUTOADV_INST, &advp,
+                                   NULL, _gap_event_cb, _gap_cb_arg);
+    assert(rc == 0);
+
+    /* get mbuf for adv data, this is freed from the `ble_gap_ext_adv_set_data` */
+    struct os_mbuf *data;
+
+    data = os_msys_get_pkthdr(BLE_HS_ADV_MAX_SZ, 0);
+    assert(data);
+
+    /* fill mbuf with adv data */
+    rc = os_mbuf_append(data, _ad.buf, _ad.pos);
+    assert(rc == 0);
+
+    /* set adv data */
+    rc = ble_gap_ext_adv_set_data(CONFIG_NIMBLE_AUTOADV_INST, data);
+    assert(rc == 0);
+
+    /* set a single advertisement event */
+    int32_t dur = (_cfg.adv_duration_ms == BLE_HS_FOREVER) ? 0
+                                                           : _cfg.adv_duration_ms / 10;
```

Comment:
At first I thought the `/ 10` was a typo, but then I've read the documentation, that "units are 10 milliseconds" (which I find a bit weird). A short comment would remove this irritation.

Reply:
Added comment